Wrong error from #903 when passing _Assume_bounds_cast directly to function with itype

In this program `itype-assume-bounds-error.c`:
```c
void foo(_Ptr<int> p);
void foo_itype(int *p : itype(_Ptr<int>));

void bar() {
  int *p = 0;
  foo(_Assume_bounds_cast<_Ptr<int>>(p));
  foo_itype(_Assume_bounds_cast<_Ptr<int>>(p));

  _Ptr<int> q = _Assume_bounds_cast<_Ptr<int>>(p);
  foo(q);
  foo_itype(q);
}
```

After #903, `clang -c -o /dev/null itype-assume-bounds-error.c` gives the following error:
```
itype-assume-bounds-error.c:7:13: error: it is not possible to prove argument meets declared bounds for 1st parameter
  foo_itype(_Assume_bounds_cast<_Ptr<int>>(p));
            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
itype-assume-bounds-error.c:7:13: note: the expected argument bounds use the variable 'p' and there is no relational information involving 'p' and any of the expressions used by the inferred bounds
itype-assume-bounds-error.c:7:13: note: (expanded) expected argument bounds are 'bounds((_Array_ptr<int>)_Assume_bounds_cast<_Ptr<int>>p(count(1)), (_Array_ptr<int>)_Assume_bounds_cast<_Ptr<int>>p(count(1)) + 1)'
itype-assume-bounds-error.c:7:13: note: (expanded) inferred bounds are 'bounds((_Array_ptr<int>)value of p, (_Array_ptr<int>)value of p + 1)'
  foo_itype(_Assume_bounds_cast<_Ptr<int>>(p));
            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
Before #903, this was a warning.

I'm no expert on the Checked C rules, but it looks like the cause of the error is that the `_Assume_bounds_cast` is present in the expression for the expected bounds but not the actual bounds.  The fact that the error doesn't occur when the casted value is assigned to another local variable before being passed to `foo_itype` leads me to believe this is a bug.  It's interesting that the problem does not occur with a function `foo` that takes the checked type outright rather than using an itype.

This is a simplified version of a failure we saw in [a recently added 3C regression test that has not yet been submitted to this repository](https://github.com/correctcomputation/checkedc-clang/blob/a17fd2de5e33595e1c1f045539f5aace15ba2abb/clang/test/3C/liberal_itypes_ptrptr.c#L147) when we tried to merge from this repository (correctcomputation/checkedc-clang#407).  I'm hoping that the root cause is the same and a fix to the case above will also fix our regression test.

--- a/clang/include/clang/AST/RecursiveASTVisitor.h
+++ b/clang/include/clang/AST/RecursiveASTVisitor.h
@@ -2617,7 +2617,6 @@ DEF_TRAVERSE_STMT(NullaryBoundsExpr, {})
 DEF_TRAVERSE_STMT(RangeBoundsExpr, {})
 DEF_TRAVERSE_STMT(InteropTypeExpr, {})
 DEF_TRAVERSE_STMT(PositionalParameterExpr, {})
-DEF_TRAVERSE_STMT(BoundsValueExpr, {})
 DEF_TRAVERSE_STMT(CHKCBindTemporaryExpr, {})
 DEF_TRAVERSE_STMT(PackExpr, {})
 
@@ -2673,6 +2672,11 @@ DEF_TRAVERSE_STMT(AsTypeExpr, {})
 // CheckedC Bounds Casting
 DEF_TRAVERSE_STMT(BoundsCastExpr, {
   TRY_TO(TraverseTypeLoc(S->getTypeInfoAsWritten()->getTypeLoc()));
+})
+
+// CheckedC Bounds Value Expressions
+DEF_TRAVERSE_STMT(BoundsValueExpr, {
+  TRY_TO(TraverseStmt(S->getTemporaryBinding()));
 })
 
 

--- a/clang/lib/Sema/SemaBounds.cpp
+++ b/clang/lib/Sema/SemaBounds.cpp
@@ -555,6 +555,15 @@ namespace {
           ++Count;
         return true;
       }
+
+      // Do not traverse the child of a BoundsValueExpr.
+      // If a BoundsValueExpr uses the variable V, this should not count
+      // toward the total occurrence count of V in the expression.
+      // For example, for the expression BoundsValue(TempBinding(v)) + v, the
+      // total occurrence count of the variable v should be 1, not 2.
+      bool TraverseBoundsValueExpr(BoundsValueExpr *E) {
+        return true;
+      }
   };
 
   // VariableOccurrenceCount returns the number of occurrences of V in E.
@@ -667,6 +676,14 @@ namespace {
         else
           return ExprCreatorUtil::CreateImplicitCast(SemaRef, Child,
                                                      CK, E->getType());
+      }
+
+      // Do not replace expressions within a BoundsValueExpr.
+      // For example, the result of replacing `v` with `v + 1` in
+      // BoundsValue(TempBinding(v)) should be BoundsValue(TempBinding(v)),
+      // not BoundsValue(TempBinding(v + 1)).
+      bool TraverseBoundsValueExpr(BoundsValueExpr *E) {
+        return true;
       }
   };
 

--- a/clang/lib/Sema/SemaBounds.cpp
+++ b/clang/lib/Sema/SemaBounds.cpp
@@ -6573,6 +6573,14 @@ namespace {
       return true;
     }
 
+    // Do not traverse the children of a BoundsValueExpr. Any expressions
+    // that are wrapped in a BoundsValueExpr should not be considered
+    // modifying expressions. For example, BoundsValue(TempBinding(f()))
+    // should not be considered modifying.
+    bool TraverseBoundsValueExpr(BoundsValueExpr *E) {
+      return true;
+    }
+
 
   private:
     Sema &S;

--- a/clang/lib/Sema/SemaBounds.cpp
+++ b/clang/lib/Sema/SemaBounds.cpp
@@ -677,14 +677,6 @@ namespace {
           return ExprCreatorUtil::CreateImplicitCast(SemaRef, Child,
                                                      CK, E->getType());
       }
-
-      // Do not replace expressions within a BoundsValueExpr.
-      // For example, the result of replacing `v` with `v + 1` in
-      // BoundsValue(TempBinding(v)) should be BoundsValue(TempBinding(v)),
-      // not BoundsValue(TempBinding(v + 1)).
-      bool TraverseBoundsValueExpr(BoundsValueExpr *E) {
-        return true;
-      }
   };
 
   // If an original value is provided, ReplaceVariableReferences returns

--- a/clang/test/CheckedC/static-checking/free-variables.c
+++ b/clang/test/CheckedC/static-checking/free-variables.c
@@ -172,3 +172,17 @@ void g(void) {
                                               // expected-note {{inferred bounds are}}
 }
 
+void f_itype(int *p : itype(ptr<int>));
+void f_arrayptr(array_ptr<int> p : count(2));
+
+void h(void) {
+  int *p = 0;
+  f_itype(_Assume_bounds_cast<ptr<int>>(p)); // expected-warning {{cannot prove argument meets declared bounds for 1st parameter}} \
+                                             // expected-note {{(expanded) expected argument bounds are 'bounds((_Array_ptr<int>)_Assume_bounds_cast<_Ptr<int>>p(count(1)), (_Array_ptr<int>)_Assume_bounds_cast<_Ptr<int>>p(count(1)) + 1)'}} \
+                                             // expected-note {{(expanded) inferred bounds are 'bounds((_Array_ptr<int>)value of p, (_Array_ptr<int>)value of p + 1)'}}
+
+  array_ptr<int> a : count(2) = 0;
+  f_arrayptr(_Assume_bounds_cast<array_ptr<int>>(a, count(2))); // expected-warning {{cannot prove argument meets declared bounds for 1st parameter}} \
+                                                                // expected-note {{(expanded) expected argument bounds are 'bounds(_Assume_bounds_cast<_Array_ptr<int>>a(count(2)), _Assume_bounds_cast<_Array_ptr<int>>a(count(2)) + 2)'}} \
+                                                                // expected-note {{(expanded) inferred bounds are 'bounds((_Array_ptr<int>)value of a, (_Array_ptr<int>)value of a + 2)'}}
+}

